File: sse_cli.js
#!/usr/bin/env node
// Node 18+
// Usage:
//   node sse-cli.js                  # uses default URL & sample payload
//   node sse-cli.js -u http://127.0.0.1:8000/invoke -f body.json
//   echo '{"anthropic_version":"bedrock-2023-05-31","max_tokens":512,"messages":[{"role":"user","content":"hi"}]}' | node sse-cli.js -u http://127.0.0.1:8000/invoke

const { readFileSync } = require("fs");

function parseArgs(argv) {
  const args = { url: "http://127.0.0.1:8000/invoke", file: null };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if ((a === "-u" || a === "--url") && argv[i+1]) { args.url = argv[++i]; continue; }
    if ((a === "-f" || a === "--file") && argv[i+1]) { args.file = argv[++i]; continue; }
  }
  return args;
}

function getStdin() {
  return new Promise((resolve) => {
    let data = "";
    if (process.stdin.isTTY) return resolve("");
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", chunk => data += chunk);
    process.stdin.on("end", () => resolve(data));
  });
}

function samplePayload() {
  return {
    anthropic_version: "bedrock-2023-05-31",
    max_tokens: 2048,
    messages: [{ role: "user", content: "what is ArgoCD" }],
  };
}

// Process one SSE event payload (string after "data: ")
function handleEventJSON(s, state) {
  try {
    const evt = JSON.parse(s);

    // Common Anthropic/Bedrock stream events
    switch (evt.type) {
      case "message_start":
        state.started = true;
        return;
      case "content_block_delta":
        if (evt.delta?.type === "text_delta" && typeof evt.delta.text === "string") {
          process.stdout.write(evt.delta.text);
        }
        return;
      case "content_block_stop":
      case "message_delta":
        return;
      case "message_stop":
        // ensure trailing newline
        process.stdout.write("\n");
        return;
      default:
        // ignore other event types; uncomment to debug:
        // console.error("[event]", evt.type);
        return;
    }
  } catch (e) {
    // Not JSON? print raw for debugging
    // console.error("[non-json data]", s);
  }
}

(async () => {
  try {
    const { url, file } = parseArgs(process.argv);

    // Build body: file > stdin > sample
    let bodyStr = "";
    if (file) {
      bodyStr = readFileSync(file, "utf8");
    } else {
      const stdin = await getStdin();
      bodyStr = stdin.trim() ? stdin : JSON.stringify(samplePayload());
    }

    const resp = await fetch(url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: bodyStr,
      // duplex enables streaming request bodies; harmless here but ok to include
      duplex: "half",
    });

    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      console.error(`[HTTP ${resp.status}] ${text}`);
      process.exit(1);
    }

    // If server is not streaming (returns a single JSON), just print it and exit
    if (!resp.body || typeof resp.body.getReader !== "function") {
      const t = await resp.text();
      console.log(t);
      return;
    }

    const reader = resp.body.getReader();
    const decoder = new TextDecoder();
    let buf = "";
    const state = { started: false };

    // SSE frames are typically separated by \n\n; each line starts with "data: "
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });

      // handle line-by-line; split conservatively on single newlines
      let idx;
      while ((idx = buf.indexOf("\n")) >= 0) {
        const line = buf.slice(0, idx).trimEnd();
        buf = buf.slice(idx + 1);

        // Only handle "data: ..." lines
        if (line.startsWith("data:")) {
          const payload = line.slice(5).trimStart();
          // Special terminator some APIs use:
          if (payload === "[DONE]") { process.stdout.write("\n"); break; }
          handleEventJSON(payload, state);
        }
      }
    }

    // Flush any trailing JSON (in case last chunk didn't end with newline)
    const trailing = buf.trim();
    if (trailing.startsWith("data:")) {
      handleEventJSON(trailing.slice(5).trimStart(), {});
    }
  } catch (err) {
    console.error("[client error]", err && err.message ? err.message : err);
    process.exit(1);
  }
})();

File: sse_cli.py
#!/usr/bin/env python3
"""
LLM CLI with live Markdown rendering using the Rich library.
- Streams Server-Sent Events (SSE-style) from a local /invoke endpoint
- Renders incremental markdown in-place (Live) as content arrives
- Shows model name as a simple separator line once known
- Auto-scrolls in real-time (no overflow dots)
- Supports aborting current stream with:
    - Pressing 'q' during streaming
    - Cmd+\\ (Mac) or Ctrl+\\ (others) signal

Requirements:
    pip install rich requests

Usage:
    python llm_cli.py --url http://127.0.0.1:8000/invoke
"""
from __future__ import annotations

import argparse
import json
import os
import select
import signal
import sys
import termios
import tty
from contextlib import contextmanager
from typing import Optional

import requests
from requests.exceptions import RequestException
from rich.console import Console
from rich.markdown import Markdown
from rich.live import Live
from rich.prompt import Prompt
from rich.text import Text

DEFAULT_URL = "http://127.0.0.1:8000/invoke"

COLOR_PROMPT = "bold green"
COLOR_MODEL = "cyan"

console = Console(force_terminal=True, force_interactive=True, highlight=False, soft_wrap=True)
_abort = False


# ---------------- Key capture helpers ----------------
@contextmanager
def raw_mode(file):
    if not file.isatty():
        yield
        return
    old_attrs = termios.tcgetattr(file.fileno())
    try:
        tty.setcbreak(file.fileno())
        yield
    finally:
        termios.tcsetattr(file.fileno(), termios.TCSADRAIN, old_attrs)


def check_for_q(timeout: float = 0.0) -> bool:
    if not sys.stdin.isatty():
        return False
    rlist, _, _ = select.select([sys.stdin], [], [], timeout)
    if rlist:
        ch = os.read(sys.stdin.fileno(), 1)
        if ch in (b"q", b"Q"):
            return True
    return False


# ---------------- Signal handlers ----------------
def _signal_handler(sig, frame):
    global _abort
    _abort = True
    console.print("\n[red]Aborted by signal[/red]")


signal.signal(signal.SIGQUIT, _signal_handler)


# ---------------- HTTP helpers ----------------
def build_payload(user_input: str) -> dict:
    return {
        "anthropic_version": "bedrock-2023-05-31",
        "max_tokens": 2048,
        "messages": [
            {"role": "user", "content": user_input}
        ],
    }


def _pretty_http_error(resp: requests.Response) -> None:
    try:
        payload = resp.json()
        msg = json.dumps(payload, indent=2)
    except Exception:
        msg = resp.text
    console.print(f"[bold red]HTTP {resp.status_code}[/bold red]\n\n{msg}")


def _event_iter_lines(resp: requests.Response):
    for raw in resp.iter_lines(decode_unicode=True):
        if not raw:
            continue
        if raw.startswith("data:"):
            yield raw[5:].lstrip()
        else:
            yield raw


# ---------------- Streaming ----------------
def stream_response(url: str, payload: dict) -> Optional[str]:
    global _abort
    _abort = False
    try:
        with requests.post(url, json=payload, stream=True, timeout=60) as r:
            if not r.ok:
                _pretty_http_error(r)
                return None

            md_text = ""
            model_name: Optional[str] = None

            with raw_mode(sys.stdin), Live(console=console, refresh_per_second=24, transient=False, auto_refresh=True, vertical_overflow="visible") as live:
                for data in _event_iter_lines(r):
                    if _abort or check_for_q(0):
                        console.print("\n[red]Aborted[/red]")
                        break
                    if data == "[DONE]":
                        break
                    try:
                        evt = json.loads(data)
                    except json.JSONDecodeError:
                        continue

                    etype = evt.get("type")
                    if etype == "message_start" and isinstance(evt.get("message"), dict):
                        model_name = evt["message"].get("model") or model_name
                        if model_name:
                            console.rule(f"[bold {COLOR_MODEL}]{model_name}")

                    if etype == "content_block_delta":
                        delta = evt.get("delta", {})
                        if delta.get("type") == "text_delta":
                            md_text += delta.get("text", "")

                    live.update(Markdown(md_text), refresh=True)

                    if etype == "message_stop":
                        break

            if not _abort:
                console.print(Markdown(md_text))
            return md_text

    except RequestException as e:
        console.print(f"[red]Network error[/red]: {e}")
        return None


# ---------------- Interactive loop ----------------
def interactive_loop(url: str) -> None:
    console.rule("LLM CLI • Live Markdown")
    console.print(Text("Type 'exit' or 'quit' to leave. Press 'q' during stream, or Cmd+\\ (Mac)/Ctrl+\\ to abort.", style="dim"))

    while True:
        try:
            user_input = Prompt.ask(Text("prompt>", style=COLOR_PROMPT)).strip()
        except (EOFError, KeyboardInterrupt):
            console.print("\n[dim]Bye![/dim]")
            return

        if not user_input:
            continue
        if user_input.lower() in {"exit", "quit"}:
            console.print("[dim]Bye![/dim]")
            return

        payload = build_payload(user_input)
        stream_response(url, payload)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Stream LLM responses and render markdown live in the console.")
    parser.add_argument("--url", default=DEFAULT_URL, help=f"Inference endpoint (default: {DEFAULT_URL})")
    args = parser.parse_args(argv)

    try:
        interactive_loop(args.url)
        return 0
    except KeyboardInterrupt:
        console.print("\n[dim]Bye![/dim]")
        return 0


if __name__ == "__main__":
    raise SystemExit(main())
